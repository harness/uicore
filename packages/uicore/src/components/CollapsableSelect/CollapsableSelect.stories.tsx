/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary } from '@storybook/addon-docs/blocks'
import { CollapsableSelectOptions, CollapsableSelectType, FormikCollapsableSelect } from './CollapsableSelect'
import { Form, Formik } from 'formik'
import { noop } from 'lodash-es'
import * as Yup from 'yup'
import { Icon } from '@harness/icons'
import { Layout, Text } from '../..'

export default {
  title: 'Components / CollapsableSelect',

  component: FormikCollapsableSelect,
  parameters: {
    layout: 'centered',
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>FormikCollapsableSelect</Title>
            <Subtitle>
              <pre>
                <code>{`import { FormikCollapsableSelect } from '@harness/uicore'`}</code>
              </pre>
            </Subtitle>

            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta

interface T {
  text: string
}

interface U {
  info: string
  icon: JSX.Element
}

export const CardView: Story<{ items: Array<CollapsableSelectOptions & T> }> = args => {
  return (
    <Formik
      initialValues={{ connectivityMode: '' }}
      onSubmit={noop}
      validationSchema={Yup.object().shape({
        connectivityMode: Yup.string().trim().required('Connectivity Mode  is required')
      })}>
      {formik => (
        <Form>
          <FormikCollapsableSelect<T>
            type={CollapsableSelectType.CardView}
            {...args}
            name={'connectivityMode'}
            selected={args.items[args.items.findIndex(item => item.value === formik.values.connectivityMode)]}
            renderItem={item => {
              return <Layout.Vertical>{item.text}</Layout.Vertical>
            }}
          />
        </Form>
      )}
    </Formik>
  )
}

export const Custom: Story<{ items: Array<CollapsableSelectOptions & U>; isReadonly: boolean }> = args => {
  return (
    <Formik
      initialValues={{ connectivityMode: '' }}
      onSubmit={noop}
      validationSchema={Yup.object().shape({
        connectivityMode: Yup.string().trim().required('Connectivity Mode  is required')
      })}>
      {formik => (
        <Form>
          <FormikCollapsableSelect<U>
            {...args}
            selected={args.items[args.items.findIndex(item => item.value === formik.values.connectivityMode)]}
            name={'connectivityMode'}
            renderItem={item => {
              return (
                <Layout.Vertical border={true} spacing="medium">
                  {item.icon}
                  <Text font="small">{item.info}</Text>
                </Layout.Vertical>
              )
            }}
          />
        </Form>
      )}
    </Formik>
  )
}

CardView.args = {
  items: [
    {
      text: 'This option is to select connectivity through Delegate',
      value: 'Delegate'
    },
    {
      text: 'This option is to select connectivity through Manager',
      value: 'Manager'
    }
  ]
}

Custom.args = {
  items: [
    {
      icon: <Icon name="service-jenkins" size={50} />,
      info: 'This option selects Jenkins',
      value: 'Jenkins'
    },
    {
      icon: <Icon name="service-gcp" size={50} />,
      info: 'This option selects Gcp',
      value: 'Gcp'
    }
  ]
}
