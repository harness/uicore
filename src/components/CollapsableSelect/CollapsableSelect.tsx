/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useEffect } from 'react'
import { FormGroup, HTMLDivProps, Intent } from '@blueprintjs/core'
import { FormikContextType, connect } from 'formik'
import cx from 'classnames'
import { isEqual, isEmpty, get } from 'lodash-es'
import { Card } from '../Card/Card'
import { Layout, LayoutProps } from '../../layouts/Layout'
import { Button } from '../Button/Button'
import { FormError } from '../FormError/FormError'
import { errorCheck } from '../../core/Utils'
import css from './CollapsableSelect.css'

export enum CollapsableSelectType {
  CardView = 'CardView'
}

export interface CollapsableSelectOptions {
  value: string | number | symbol
}

export interface CollapsableSelectProps<T> extends Omit<HTMLDivProps, 'onChange'> {
  selected: (CollapsableSelectOptions & T) | undefined
  items: Array<CollapsableSelectOptions & T>
  itemClassName?: string
  renderItem: (item: T, selected?: CollapsableSelectOptions & T) => JSX.Element
  onChange?: (selected: CollapsableSelectOptions & T, e: React.MouseEvent<HTMLDivElement>) => void
  layoutProps?: Partial<LayoutProps>
  name: string
  isReadonly?: boolean
  changeText?: string
  cancelText?: string
  type?: CollapsableSelectType
  actionBtnClassName?: string
}

export interface ConnectedCollapsableSelectProps<T> extends CollapsableSelectProps<T> {
  formik: FormikContextType<Record<string, unknown>>
}
export function CollapsableSelect<T>(props: ConnectedCollapsableSelectProps<T>) {
  const {
    className,
    itemClassName,
    actionBtnClassName,
    formik,
    name,
    items = [],
    selected,
    renderItem,
    changeText = 'Change',
    cancelText = 'Close',
    layoutProps,
    isReadonly = false,
    type = undefined
  } = props

  const value = get(formik?.values, name)

  const [showAllOptions, setShowAllOptions] = React.useState(!isEmpty(value))

  const hasError = errorCheck(name, formik)
  const intent = hasError ? Intent.DANGER : Intent.NONE
  const helperText = hasError ? <FormError errorMessage={get(formik?.errors, name)} name={name} /> : null

  function handleChangeClick(): void {
    setShowAllOptions(true)
  }

  function handleCancelClick(): void {
    setShowAllOptions(false)
  }

  function handleChange(item: CollapsableSelectOptions & T, event: React.MouseEvent<HTMLDivElement>): void {
    formik.setFieldValue(name, item.value)
    formik.setFieldTouched(name, true, false)
    props.onChange?.(item, event)
  }

  useEffect(() => {
    setShowAllOptions(isEmpty(value))
  }, [value])

  const selectedItemIndex = value ? items.findIndex(item => item.value === value) : -1
  let visibleItems =
    selectedItemIndex > -1
      ? [items[selectedItemIndex], ...items.filter((_val, index) => index !== selectedItemIndex)]
      : items

  if (!showAllOptions) {
    visibleItems = [visibleItems[0]]
  }

  return (
    <FormGroup className={className} helperText={helperText} intent={intent}>
      <Layout.Horizontal spacing={'medium'} {...layoutProps}>
        {visibleItems.map((item, index) => {
          const isSelected = isEqual(item, selected)

          return (
            <React.Fragment key={index}>
              {type === CollapsableSelectType.CardView ? (
                <Card
                  className={itemClassName}
                  interactive={true}
                  data-index={index}
                  selected={isSelected}
                  cornerSelected={true}
                  onClick={event => handleChange(item, event)}>
                  {renderItem(item, selected)}
                </Card>
              ) : (
                <div className={itemClassName} data-index={index} onClick={event => handleChange(item, event)}>
                  {renderItem(item, selected)}
                </div>
              )}
            </React.Fragment>
          )
        })}

        {showAllOptions ? null : (
          <Button
            className={cx(css.closeChangeBtn, actionBtnClassName)}
            disabled={isReadonly}
            minimal
            icon="Edit"
            iconProps={{ size: 12 }}
            withoutCurrentColor={true}
            intent="primary"
            data-testid="collapsable-select-change"
            onClick={handleChangeClick}
            text={changeText}
          />
        )}
        {showAllOptions && value ? (
          <Button
            className={cx(css.closeChangeBtn, actionBtnClassName)}
            disabled={isReadonly}
            minimal
            icon="cross"
            iconProps={{ size: 12 }}
            withoutCurrentColor={true}
            intent="primary"
            data-testid="thumbnail-select-cancel"
            onClick={handleCancelClick}
            text={cancelText}
          />
        ) : null}
      </Layout.Horizontal>
    </FormGroup>
  )
}

export const FormikCollapsableSelect = connect(CollapsableSelect) as <T>(
  props: CollapsableSelectProps<T>
) => React.ReactElement<CollapsableSelectProps<T>>
