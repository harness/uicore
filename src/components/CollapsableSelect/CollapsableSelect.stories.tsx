import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { Title, Subtitle, ArgsTable, Stories, PRIMARY_STORY, Primary } from '@storybook/addon-docs/blocks'
import { CollapsableSelectOptions, CollapsableSelectType, FormikCollapsableSelect } from './CollapsableSelect'
import { Formik } from 'formik'
import { noop } from 'lodash-es'
import * as Yup from 'yup'
import { Icon, Layout } from 'index'

export default {
  title: 'Components / CollapsableSelect',

  component: FormikCollapsableSelect,
  parameters: {
    docs: {
      source: {
        type: 'code'
      },

      page: function PageDescription() {
        return (
          <>
            <Title>FormikCollapsableSelect</Title>
            <Subtitle>
              <pre>
                <code>{`import { FormikCollapsableSelect } from '@wings-software/uicore'`}</code>
              </pre>
            </Subtitle>

            <Primary />
            <ArgsTable story={PRIMARY_STORY} />

            <Stories />
          </>
        )
      }
    }
  },
  decorators: [Story => <Story />]
} as Meta

interface T {
    text: string
}

interface U {
    icon: JSX.Element
}

export const CardView: Story<{ items: Array<CollapsableSelectOptions & T>; }> = args => {
  return (
    <Formik
      initialValues={{ connectivityMode: '' }}
      onSubmit={noop}
      validationSchema={Yup.object().shape({
        connectivityMode: Yup.string().trim().required('Connectivity Mode  is required')
      })}>
      <FormikCollapsableSelect<T> type={CollapsableSelectType.CardView} {...args} name={'connectivityMode'} renderItem={(item)=>{
          <Layout.Vertical>{item.text}</Layout.Vertical>
      }} />
    </Formik>
  )
}

export const Custom: Story<{ items:  Array<CollapsableSelectOptions & U>; isReadonly: boolean }> = args => {
  return (
    <Formik
    initialValues={{ connectivityMode: '' }}
    onSubmit={noop}
    validationSchema={Yup.object().shape({
      connectivityMode: Yup.string().trim().required('Connectivity Mode  is required')
    })}>
      <FormikCollapsableSelect<U>  {...args} name={'connectivityMode'}  renderItem={(item)=>{
          item.icon            
      }}/>
    </Formik>
  )
}

CardView.args = {
    items: [
      {
        text : 'This option is to select connectivity through Delegate',
        value: 'Delegate'
      },
      {
        text : 'This option is to select connectivity through Manager',
        value: 'Manager'
      }]
    }

 Custom.args = {
        items: [
          {
            icon: <Icon  name='service-jenkins' />,
      
            value: 'Jenkins'
          },
          {
            icon: <Icon  name='service-gcp' />,
            value: 'Gcp'
          }]
         }  