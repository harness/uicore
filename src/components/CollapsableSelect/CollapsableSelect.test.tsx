/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, fireEvent, act } from '@testing-library/react'
import { CollapsableSelectType, FormikCollapsableSelect } from './CollapsableSelect'
import { Form, Formik } from 'formik'
import { Layout } from 'index'
import { noop } from 'lodash'

const items = [
  {
    text: 'Kubernetes',
    value: 'service-kubernetes',
    icon: 'service-kubernetes'
  },
  {
    text: 'Github',
    value: 'service-github',
    icon: 'service-github'
  }
]

describe('Test render CollapsableSelect', () => {
  test('should render card view with defaultValue', () => {
    const { container, getByText } = render(
      <Formik initialValues={{ connectivityMode: 'service-kubernetes' }} onSubmit={jest.fn()}>
        {formik => (
          <Form>
            <FormikCollapsableSelect
              type={CollapsableSelectType.CardView}
              items={items}
              name={'connectivityMode'}
              selected={items[items.findIndex(item => item.value === formik.values.connectivityMode)]}
              renderItem={item => {
                return <Layout.Vertical>{item.text}</Layout.Vertical>
              }}
            />
          </Form>
        )}
      </Formik>
    )
    expect(getByText('Change')).toBeInTheDocument()
    expect(container.querySelector('[data-icon="main-tick"]')).not.toBeNull()

    expect(container).toMatchSnapshot()
  })

  test('should render card view with no defaultValue', () => {
    const { container, getByText } = render(
      <Formik initialValues={{ connectivityMode: '' }} onSubmit={noop}>
        {formik => (
          <Form>
            <FormikCollapsableSelect
              type={CollapsableSelectType.CardView}
              items={items}
              name={'connectivityMode'}
              selected={undefined}
              onChange={() => {
                formik.setFieldTouched('connectivityMode', true)
              }}
              renderItem={item => {
                return <Layout.Vertical>{item?.text}</Layout.Vertical>
              }}
            />
          </Form>
        )}
      </Formik>
    )
    // displays other options as well
    expect(getByText('Github')).toBeInTheDocument()
    // nothing selected
    expect(container.querySelector('[data-icon="main-tick"]')).toBeNull()
    expect(container).toMatchSnapshot()
  })

  test('collapsable select flow ', () => {
    const { container, getByText } = render(
      <Formik initialValues={{ connectivityMode: 'service-kubernetes' }} onSubmit={noop}>
        {formik => (
          <Form>
            <FormikCollapsableSelect
              type={CollapsableSelectType.CardView}
              items={items}
              name={'connectivityMode'}
              selected={items[items.findIndex(item => item.value === formik.values.connectivityMode)]}
              renderItem={item => {
                return <Layout.Vertical>{item?.text}</Layout.Vertical>
              }}
            />
          </Form>
        )}
      </Formik>
    )
    expect(container.querySelector('[data-icon="main-tick"]')).not.toBeNull()

    act(() => {
      fireEvent.click(getByText('Change'))
    })
    // Close and other options appear
    expect(getByText('Github')).toBeInTheDocument()
    expect(getByText('Close')).toBeInTheDocument()

    const secondOption = container.querySelector('[data-index="1"]')

    act(() => {
      if (secondOption) fireEvent.click(secondOption)
    })

    expect(getByText('Change')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('should render customView ', () => {
    const { container } = render(
      <Formik initialValues={{ connectivityMode: 'service-kubernetes' }} onSubmit={noop}>
        {formik => (
          <Form>
            <FormikCollapsableSelect
              items={items}
              name={'connectivityMode'}
              selected={items[items.findIndex(item => item.value === formik.values.connectivityMode)]}
              renderItem={item => {
                return <Layout.Vertical>{item?.text}</Layout.Vertical>
              }}
            />
          </Form>
        )}
      </Formik>
    )
    // View will not have card  and will be controlled by user
    expect(container.querySelector('[class="bp3-card bp3-interactive bp3-elevation-0 card interactive"]')).toBeNull()

    expect(container).toMatchSnapshot()
  })
})
