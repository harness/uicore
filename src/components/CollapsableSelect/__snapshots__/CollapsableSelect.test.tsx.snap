// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render CollapsableSelect collapsable select flow  1`] = `
<div>
  <form>
    <div
      class="bp3-form-group"
    >
      <div
        class="bp3-form-content"
      >
        <div
          class="horizontal layout-spacing-medium main spacing spacing-medium"
        >
          <div
            class="bp3-card bp3-interactive bp3-elevation-0 card selected interactive"
            data-index="0"
            tabindex="0"
          >
            <span
              class="corner"
            >
              <span
                class="badge"
              />
              <span
                class="bp3-icon main color color-white"
                data-icon="main-tick"
              >
                <svg
                  height="10"
                  width="10"
                />
              </span>
            </span>
            <div
              class="vertical layout-spacing-undefined main"
            >
              Github
            </div>
          </div>
          <button
            class="bp3-button bp3-minimal button font main closeChangeBtn intent intent-primary without-shadow withLeftIcon"
            data-testid="collapsable-select-change"
            type="button"
          >
            <span
              class="bp3-icon main padding padding-right padding-right-xsmall"
              data-icon="Edit"
            >
              <svg
                height="12"
                width="12"
              />
            </span>
            <span
              class="bp3-button-text"
            >
              Change
            </span>
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Test render CollapsableSelect should render card view with defaultValue 1`] = `
<div>
  <form>
    <div
      class="bp3-form-group"
    >
      <div
        class="bp3-form-content"
      >
        <div
          class="horizontal layout-spacing-medium main spacing spacing-medium"
        >
          <div
            class="bp3-card bp3-interactive bp3-elevation-0 card selected interactive"
            data-index="0"
            tabindex="0"
          >
            <span
              class="corner"
            >
              <span
                class="badge"
              />
              <span
                class="bp3-icon main color color-white"
                data-icon="main-tick"
              >
                <svg
                  height="10"
                  width="10"
                />
              </span>
            </span>
            <div
              class="vertical layout-spacing-undefined main"
            >
              Kubernetes
            </div>
          </div>
          <button
            class="bp3-button bp3-minimal button font main closeChangeBtn intent intent-primary without-shadow withLeftIcon"
            data-testid="collapsable-select-change"
            type="button"
          >
            <span
              class="bp3-icon main padding padding-right padding-right-xsmall"
              data-icon="Edit"
            >
              <svg
                height="12"
                width="12"
              />
            </span>
            <span
              class="bp3-button-text"
            >
              Change
            </span>
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Test render CollapsableSelect should render card view with no defaultValue 1`] = `
<div>
  <form>
    <div
      class="bp3-form-group"
    >
      <div
        class="bp3-form-content"
      >
        <div
          class="horizontal layout-spacing-medium main spacing spacing-medium"
        >
          <div
            class="bp3-card bp3-interactive bp3-elevation-0 card interactive"
            data-index="0"
            tabindex="0"
          >
            <div
              class="vertical layout-spacing-undefined main"
            >
              Kubernetes
            </div>
          </div>
          <div
            class="bp3-card bp3-interactive bp3-elevation-0 card interactive"
            data-index="1"
            tabindex="0"
          >
            <div
              class="vertical layout-spacing-undefined main"
            >
              Github
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Test render CollapsableSelect should render customView  1`] = `
<div>
  <form>
    <div
      class="bp3-form-group"
    >
      <div
        class="bp3-form-content"
      >
        <div
          class="horizontal layout-spacing-medium main spacing spacing-medium"
        >
          <div
            data-index="0"
          >
            <div
              class="vertical layout-spacing-undefined main"
            >
              Kubernetes
            </div>
          </div>
          <button
            class="bp3-button bp3-minimal button font main closeChangeBtn intent intent-primary without-shadow withLeftIcon"
            data-testid="collapsable-select-change"
            type="button"
          >
            <span
              class="bp3-icon main padding padding-right padding-right-xsmall"
              data-icon="Edit"
            >
              <svg
                height="12"
                width="12"
              />
            </span>
            <span
              class="bp3-button-text"
            >
              Change
            </span>
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
